### PR TITLE
Server config

### LINKED ISSUE #17: Load server network configuration from files
Implement some sort of config that we can load the ip/port from (and possibly other stuff as we go along).

### PR DESCRIPTION
**Pull request by liamdalg**
Adds JSON config in the file server-config.json. A config is auto-generated if one can't be loaded.

Resolves #17

### CODE REVIEW

File: ProjectFiles/Assets/Scripts/Network/ServerConfig.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace Network
```
Comment (RileyEv) [inline ProjectFiles/Assets/Scripts/Network/ServerConfig.cs:5]: Not sure if this should be in the network namespace? At the moment its only storing things about the network but we could end up expanding it further in the future

File: ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs [hunk 2]
```diff
@@ -9,17 +12,47 @@ public class ServerGameLoop : IGameLoop
...
 
             // Start server
-            server = new Server(world);
+            server = new Server(world, config);
             var success = server.Start();
             
             return success;
         }
 
+        private ServerConfig LoadConfigOrDefault(string path)
```
Comment (RileyEv) [inline ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs]: Should this be a static method in the `ServerConfig` class? I feel like this should be abstracted from the game loop.

File: ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs [hunk 3]
```diff
@@ -9,17 +12,47 @@ public class ServerGameLoop : IGameLoop
...
+            catch (ArgumentException e)
+            {
+                Debug.Log(e);
+                Debug.Log($"Failed to load {path}. Generating and running default config at server-config.json...");
+                SaveConfig(path, ServerConfig.DefaultConfig);
+                return ServerConfig.DefaultConfig;
+            }
+        }
+
+        private void SaveConfig(string path, ServerConfig config)
```
Comment (RileyEv) [inline ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs]: Same with this one?

File: ProjectFiles/Assets/Scripts/Network/Server.cs [hunk 4]
```diff
@@ -17,34 +17,34 @@ public class Server
         private World world;
 
         private Dictionary<int, int> connectionPlayerIDs;
+
+        public ServerConfig Config { get; private set; }
+        private string IP => Config.IpAddress;
+        private ushort Port => Config.Port;
```
Comment (RileyEv) [inline ProjectFiles/Assets/Scripts/Network/Server.cs:23]: What do these two things do? Is it like a symlink kinda thing?

**Review by RileyEv — COMMENTED** [review]
Added a few things we should probably think about.

File: ProjectFiles/Assets/Scripts/Network/Server.cs (hunk 4 above)
Comment (liamdalg) [reply ProjectFiles/Assets/Scripts/Network/Server.cs:23]: it's like a shorthand for a getter so you don't have to do Config.Port/Config.IpAddress. Here it is basically acting as a symlink

File: ProjectFiles/Assets/Scripts/Network/ServerConfig.cs (hunk 1 above)
Comment (ghost) [reply ProjectFiles/Assets/Scripts/Network/ServerConfig.cs:5]: Maybe just something to bear in mind for the future

File: ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs (hunk 2 above)
Comment (ghost) [reply ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs]: I think I agree but maybe Liam had a reason for this

Comment (liamdalg) [reply ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs]: Yeah I agree, I'll make some changes now :)

File: ProjectFiles/Assets/Scripts/Network/Server.cs (hunk 4 above)
Comment (ghost) [reply ProjectFiles/Assets/Scripts/Network/Server.cs:23]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members

Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Network/Server.cs:23]: I'll just think of it as a symlink for now 🤷‍♂

File: ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs (hunk 2 above)
Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs]: Okay cool :)

**Review by ghost — APPROVED** [review]

**Pull request merged**